### PR TITLE
AVRO-3756: [csharp] Add a method to return types instead of writing them to disk

### PR DESCRIPTION
**Pull request by alexrosenfeld10**
## What is the purpose of the change

This PR adds the ability to write types directly back to the user, instead of forcing writing to disk. This is critical for use in scenarios where the code generation is done in a CI job, where disk permissions might not be flexible.

## Verifying this change

This change can be verified as follows:

- Manually verified the change by building and using the new methods locally in C# Source Generators (see https://github.com/alexrosenfeld10/avro-sourcegen-csharp/pull/4)

Note - I did not add unit tests for this change nor refactor any code style, as it is a very congruent method to other methods in the `CodeGen` class. It may make sense to refactor the class to be a bit DRY-er, and at that point, add tests. If that should happen, it should happen in a separate PR.

## Documentation

- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? C# XML docs. This change is not exposed by the `avrogen` tool because in that context it doesn't really make sense to do anything _other_ than write directly to disk. This is meant for use from other C# code.

### CODE REVIEW

**Comment by alexrosenfeld10** [comment]
@RyanSkraba When you have time, please review

File: lang/csharp/src/apache/main/CodeGen/CodeGen.cs [hunk 1]
```diff
@@ -1136,6 +1136,48 @@
             }
         }
 
+        /// <summary>
+        /// Gets names and generated code of the schema(s) types
+        /// </summary>
+        /// <returns></returns>
+        public virtual IDictionary<string, string> GetTypes()
+        {
+            var cscp = new CSharpCodeProvider();
```
[the remark this answers is not in the document]
Comment (alexrosenfeld10) [reply lang/csharp/src/apache/main/CodeGen/CodeGen.cs:1145]: just copying patterns from the other (very similar) methods. If desired, I can update this.

**Comment by alexrosenfeld10** [comment]
@RyanSkraba @zcsizmadia @martin-g @KyleSchoonover I found your names from recent PRs, would someone be available to push this over the line? Thanks very much in adv.

**Comment by RyanSkraba** [comment]
Hey -- thanks for the contribution!  I've had limited time (and I've been pre-occupied with the 1.11.2 release prep).

I'm not a C# dev, but it _seems_ like this is a pretty safe change and could be included safely in the minor release.  If any C# committers can take a look and confirm, that would be great -- but any C# dev's review is valued!

**Comment by alexrosenfeld10** [comment]
Thanks @RyanSkraba , I appreciate the response. 

Yes, this is a super benign change - it's purely additive, and has no side effects. This sets up the stage for something larger I and @alfhv are working on - getting the C# side of Avro able to make cross-file schema references by namespace.

File: lang/csharp/src/apache/main/CodeGen/CodeGen.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexrosenfeld10) [reply lang/csharp/src/apache/main/CodeGen/CodeGen.cs:1145]: Sure, I can do that. My approach with OSS PRs is to match existing patterns and introduce as little deviation / other changes as possible to achieve the feature. Happy to add `using` here

[the remark this answers is not in the document]
Comment (zcsizmadia) [reply lang/csharp/src/apache/main/CodeGen/CodeGen.cs:1145]: I have no issue if you add "using" tin this PR where those other places are. new CSharpCodeProvider is used without using. (if not too many). If too many lets make another PR

Not having "using" IMO is considered a resource leak bug, so breaking the pattern in such case is poreferred.

[the remark this answers is not in the document]
Comment (alexrosenfeld10) [reply lang/csharp/src/apache/main/CodeGen/CodeGen.cs:1145]: @zcsizmadia I looked at this, it requires changing the `LangVersion` for the project.
Screenshot 2023-06-13 at 2 38 22 PM

Happy to make these changes, but it seems out of scope for this PR.

**Review by alexrosenfeld10 — COMMENTED** [review]
> I always like top get some basic test coverage to set the minimal baseline. Those initial unit tests can be removed or rewritten when a better rework is happening in the future

@zcsizmadia are you requesting I add unit tests for this? I took a look at this, there is no existing pattern for mocking test data or unit testing any similarly structured methods in the `CodeGen` class. I think defining those patterns should be something done in a separate initiative

File: lang/csharp/src/apache/main/CodeGen/CodeGen.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zcsizmadia) [reply lang/csharp/src/apache/main/CodeGen/CodeGen.cs:1145]: That using format is only C# 8+. You need to use using with brackets. Which sepcifies the scope explicitly.
image

[the remark this answers is not in the document]
Comment (alexrosenfeld10) [reply lang/csharp/src/apache/main/CodeGen/CodeGen.cs:1145]: oh duh. I haven't been in old C# in a while.. apologies

**Comment by zcsizmadia** [comment]
@alexrosenfeld10 I tthink one simple use case of what you intend to use this should be enough. I assume you will create a CodeGen object, add some schemas and call GetTypes. That output should be checked for expected result. I agree that the pattern  can be fine tuned later.

**Review by zcsizmadia — APPROVED** [review]

**Comment by zcsizmadia** [comment]
Thanks @alexrosenfeld10 !

**Comment by RyanSkraba** [comment]
Cherry-picked to [branch-1.11](https://github.com/apache/avro/commit/aa450f35caf4ed209018bf3a70fb9352989b936c).

**Comment by alexrosenfeld10** [comment]
Thanks @RyanSkraba

**Pull request merged**